File: src/main/java/cspfj/generator/constraint/LexLeqGenerator.java
package cspfj.generator.constraint;

import java.util.List;

import com.google.common.collect.Iterables;
import com.google.common.collect.Lists;

import cspfj.constraint.semantic.LexLeq;
import cspfj.exception.FailedGenerationException;
import cspfj.problem.Problem;
import cspfj.problem.Variable;
import cspom.constraint.CSPOMConstraint;

public final class LexLeqGenerator extends AbstractGenerator {

    public LexLeqGenerator(final Problem problem) {
        super(problem);
    }

    @Override
    public boolean generate(final CSPOMConstraint constraint)
            throws FailedGenerationException {

        final List<Variable> solverVariables = Lists.transform(
                constraint.getScope(), CSPOM_TO_CSP4J);

        if (Iterables.any(solverVariables, NULL_DOMAIN)) {
            return false;
        }

        addConstraint(new LexLeq(
                solverVariables.toArray(new Variable[solverVariables.size()])));
        return true;

    }

}

File: src/main/scala/cspfj/generator/constraint/LexLeqGenerator.scala
package cspfj.generator.constraint;

import cspfj.{ Variable, Problem }
import cspom.constraint.{ GeneralConstraint, CSPOMConstraint }
import cspfj.constraint.semantic.{ AllDifferent2C, AllDifferentBC }
import cspfj.constraint.semantic.ZeroSum
import cspfj.constraint.semantic.LexLeq

final class LexLeqGenerator(problem: Problem) extends AbstractGenerator(problem) {
  override def generateGeneral(constraint: GeneralConstraint) = {
    val solverVariables = constraint.scope map cspom2cspfj toArray

    if (solverVariables exists { _.dom.undefined }) {
      false
    } else {
      val (s1, s2) = solverVariables.splitAt(solverVariables.length / 2)
      require(s1.length == s2.length)
      val c = new LexLeq(s1, s2)
      //println(c)
      addConstraint(c);
      true;
    }
  }

}
